scipy/stats/stats.py

```diff
@@ -6183,71 +6284,197 @@ def ks_2samp(data1, data2, alternative='two-sided', mode='auto'):
-                    else:
-                        num_paths = _count_paths_outside_method(n1, n2, g, h)
-                        bin = special.binom(n1 + n2, n1)
-                        if not np.isfinite(bin) or not np.isfinite(num_paths) or num_paths > bin:
-                            raise FloatingPointError()
-                        prob = num_paths / bin
-
-            except FloatingPointError:
-                # Switch mode
-                mode = 'asymp'
-                saw_fp_error = True
-                # Can't raise warning here, inside the try
-            finally:
-                if saw_fp_error:
-                    if original_mode == 'exact':
-                        warnings.warn(
-                            "ks_2samp: Exact calculation overflowed. "
-                            "Switching to mode=%s" % mode, RuntimeWarning)
-                else:
-                    if prob > 1 or prob < 0:
-                        mode = 'asymp'
-                        if original_mode == 'exact':
-                            warnings.warn(
-                                "ks_2samp: Exact calculation incurred large"
-                                " rounding error. Switching to mode=%s" % mode,
-                                RuntimeWarning)
+        success, d, prob = _attempt_exact_2kssamp(n1, n2, g, d, alternative)
+        if not success:
+            mode = 'asymp'
+            if original_mode == 'exact':
+                warnings.warn(f"ks_2samp: Exact calculation unsuccessful. "
+                              f"Switching to mode={mode}.", RuntimeWarning)
 
     if mode == 'asymp':
         # The product n1*n2 is large.  Use Smirnov's asymptoptic formula.
         if alternative == 'two-sided':
-            en = np.sqrt(n1 * n2 / (n1 + n2))
-            # Switch to using kstwo.sf() when it becomes available.
-            # prob = distributions.kstwo.sf(d, int(np.round(en)))
-            prob = distributions.kstwobign.sf(en * d)
+            en = n1 * n2 / (n1 + n2)
+            prob = distributions.kstwo.sf(d, np.round(en))
         else:
             m, n = max(n1, n2), min(n1, n2)
             z = np.sqrt(m*n/(m+n)) * d
             # Use Hodges' suggested approximation Eqn 5.3
             expt = -2 * z**2 - 2 * z * (m + 2*n)/np.sqrt(m*n*(m+n))/3.0
             prob = np.exp(expt)
 
-    prob = (0 if prob < 0 else (1 if prob > 1 else prob))
-    return Ks_2sampResult(d, prob)
+    prob = np.clip(prob, 0, 1)
+    return KstestResult(d, prob)
+
+
+def _parse_kstest_args(data1, data2, args, N):
+    # kstest allows many different variations of arguments.
+    # Pull out the parsing into a separate function
+    # (xvals, yvals, )  # 2sample
+    # (xvals, cdf function,..)
+    # (xvals, name of distribution, ...)
+    # (name of distribution, name of distribution, ...)
+    # (name of distribution, empty string, ...)
+    # (name of distribution, False, ...)
+
+    # Returns xvals, yvals, cdf
```

Comment:
which combinations are possible for xvals?

Reply:
xvals is array_like.  Updated documentation.